File: mapbox_maps_flutter_mobile/ios/mapbox_maps_flutter_mobile/Sources/mapbox_maps_flutter_mobile/Classes/CameraController.swift
import MapboxMaps
import Foundation
import Flutter
import Turf

final class CameraController: _CameraManager {
    private static let errorCode = "0"

    func cameraForCoordinatesPadding(coordinates: [Point], camera: CameraOptions, coordinatesPadding: MbxEdgeInsets?, maxZoom: Double?, offset: ScreenCoordinate?) throws -> CameraOptions {
        do {
            let camera = try mapboxMap.camera(
                for: coordinates.map(\.coordinates),
                camera: camera.toCameraOptions(),
                coordinatesPadding: coordinatesPadding?.toUIEdgeInsets(),
                maxZoom: maxZoom,
                offset: offset?.toCGPoint())
            return camera.toFLTCameraOptions()
        } catch let cameraError {
            throw FlutterError(code: CameraController.errorCode, message: cameraError.localizedDescription, details: nil)
        }
    }

    func cameraForCoordinateBounds(bounds: CoordinateBounds, padding: MbxEdgeInsets?, bearing: Double?, pitch: Double?, maxZoom: Double?, offset: ScreenCoordinate?) throws -> CameraOptions {
        let camera = mapboxMap.camera(
            for: bounds.toCoordinateBounds(),
            padding: padding?.toUIEdgeInsets(),
            bearing: bearing,
            pitch: pitch,
            maxZoom: maxZoom,
            offset: offset?.toCGPoint())
        return camera.toFLTCameraOptions()
    }

    func cameraForCoordinates(coordinates: [Point], padding: MbxEdgeInsets?, bearing: Double?, pitch: Double?) throws -> CameraOptions {
        let cameraOptions = mapboxMap.camera(
            for: coordinates.map(\.coordinates),
            padding: padding?.toUIEdgeInsets(),
            bearing: bearing,
            pitch: pitch)
        return cameraOptions.toFLTCameraOptions()
    }

    func cameraForCoordinatesCameraOptions(coordinates: [Point], camera: CameraOptions, box: ScreenBox) throws -> CameraOptions {
        let cameraOptions = mapboxMap.camera(
            for: coordinates.map(\.coordinates),
            camera: camera.toCameraOptions(),
            rect: box.toCGRect())
        return cameraOptions.toFLTCameraOptions()
    }

    func cameraForGeometry(geometry: [String?: Any?], padding: MbxEdgeInsets, bearing: Double?, pitch: Double?) throws -> CameraOptions {
        let cameraOptions = mapboxMap.camera(
            for: convertDictionaryToGeometry(dict: geometry)!,
            padding: padding.toUIEdgeInsets(),
            bearing: bearing?.CGFloat,
            pitch: pitch?.CGFloat)
        return cameraOptions.toFLTCameraOptions()
    }

    func coordinateBoundsForCamera(camera: CameraOptions) throws -> CoordinateBounds {
        let bounds = self.mapboxMap.coordinateBounds(for: camera.toCameraOptions())
        return bounds.toFLTCoordinateBounds()
    }

    func coordinateBoundsForCameraUnwrapped(camera: CameraOptions) throws -> CoordinateBounds {
        // TODO use unwrapped api
        let bounds = self.mapboxMap.coordinateBounds(for: camera.toCameraOptions())
        return bounds.toFLTCoordinateBounds()
    }

    func coordinateBoundsZoomForCamera(camera: CameraOptions) throws -> CoordinateBoundsZoom {
        let coordinate = self.mapboxMap.coordinateBoundsZoom(for: camera.toCameraOptions())
        return coordinate.toFLTCoordinateBoundsZoom()
    }

    func coordinateBoundsZoomForCameraUnwrapped(camera: CameraOptions) throws -> CoordinateBoundsZoom {
        let coordinate = self.mapboxMap.coordinateBoundsZoomUnwrapped(for: camera.toCameraOptions())
        return coordinate.toFLTCoordinateBoundsZoom()
    }

    func pixelForCoordinate(coordinate: Point) throws -> ScreenCoordinate {
        let point = self.mapboxMap.point(for: coordinate.coordinates)
        return point.toFLTScreenCoordinate()
    }

    func coordinateForPixel(pixel: ScreenCoordinate) throws -> Point {
        let coordinate = self.mapboxMap.coordinate(for: pixel.toCGPoint())
        return Point(coordinate)
    }

    func pixelsForCoordinates(coordinates: [Point]) throws -> [ScreenCoordinate?] {
        let points = self.mapboxMap.points(for: coordinates.map(\.coordinates))
        return points.map({$0.toFLTScreenCoordinate()})
    }

    func coordinatesForPixels(pixels: [ScreenCoordinate?]) throws -> [Point] {
        let coordinates = self.mapboxMap.coordinates(for: pixels.compactMap({$0?.toCGPoint()}) )
        return coordinates.map(Point.init)
    }

    func setCamera(cameraOptions: CameraOptions) throws {
        self.mapboxMap.setCamera(to: cameraOptions.toCameraOptions())
    }

    func getCameraState() throws -> CameraState {
        return mapboxMap.cameraState.toFLTCameraState()
    }

    func setBounds(options: CameraBoundsOptions) throws {
        do {
            try self.mapboxMap.setCameraBounds(with: options.toCameraBoundsOptions())
        } catch {
            print("Failed to setCameraBounds. Error: \(error)")
        }
    }

    func getBounds() throws -> CameraBounds {
        return self.mapboxMap.cameraBounds.toFLTCameraBounds()
    }

    private var mapboxMap: MapboxMap
    init(withMapboxMap mapboxMap: MapboxMap) {
        self.mapboxMap = mapboxMap
    }
}

File: mapbox_maps_flutter_mobile/android/src/main/kotlin/com/mapbox/maps/mapbox_maps/CameraController.kt
package com.mapbox.maps.mapbox_maps

import android.content.Context
import com.mapbox.geojson.Point
import com.mapbox.maps.MapboxMap
import com.mapbox.maps.mapbox_maps.pigeons.*
import com.mapbox.maps.plugin.animation.*

class CameraController(private val mapboxMap: MapboxMap, private val context: Context) : _CameraManager {
  override fun cameraForCoordinatesPadding(
    coordinates: List<Point>,
    camera: CameraOptions,
    coordinatesPadding: MbxEdgeInsets?,
    maxZoom: Double?,
    offset: ScreenCoordinate?
  ): CameraOptions {
    val cameraOptions = mapboxMap.cameraForCoordinates(
      coordinates,
      camera.toCameraOptions(context),
      coordinatesPadding?.toEdgeInsets(context),
      maxZoom,
      offset?.toScreenCoordinate(context)
    )
    return cameraOptions.toFLTCameraOptions(context)
  }

  override fun cameraForCoordinateBounds(
    bounds: CoordinateBounds,
    padding: MbxEdgeInsets?,
    bearing: Double?,
    pitch: Double?,
    maxZoom: Double?,
    offset: ScreenCoordinate?
  ): CameraOptions {
    val cameraOptions = mapboxMap.cameraForCoordinateBounds(
      bounds.toCoordinateBounds(),
      padding?.toEdgeInsets(context),
      bearing,
      pitch,
      maxZoom,
      offset?.toScreenCoordinate(context)
    )
    return cameraOptions.toFLTCameraOptions(context)
  }

  override fun cameraForCoordinates(
    coordinates: List<Point>,
    padding: MbxEdgeInsets?,
    bearing: Double?,
    pitch: Double?
  ): CameraOptions {
    val cameraOptions = mapboxMap.cameraForCoordinates(
      coordinates,
      padding?.toEdgeInsets(context),
      bearing,
      pitch
    )
    return cameraOptions.toFLTCameraOptions(context)
  }

  override fun cameraForCoordinatesCameraOptions(
    coordinates: List<Point>,
    camera: CameraOptions,
    box: ScreenBox
  ): CameraOptions {
    val cameraOptions = mapboxMap.cameraForCoordinates(
      coordinates,
      camera.toCameraOptions(context),
      box.toScreenBox(context)
    )
    return cameraOptions.toFLTCameraOptions(context)
  }

  override fun cameraForGeometry(
    geometry: Map<String?, Any?>,
    padding: MbxEdgeInsets,
    bearing: Double?,
    pitch: Double?
  ): CameraOptions {
    val cameraOptions = mapboxMap.cameraForGeometry(
      geometry.toGeometry(),
      padding.toEdgeInsets(context),
      bearing,
      pitch
    )
    return cameraOptions.toFLTCameraOptions(context)
  }

  override fun coordinateBoundsForCamera(camera: CameraOptions): CoordinateBounds {
    val coordinateBounds = mapboxMap.coordinateBoundsForCamera(camera.toCameraOptions(context))
    return coordinateBounds.toFLTCoordinateBounds()
  }

  override fun coordinateBoundsForCameraUnwrapped(camera: CameraOptions): CoordinateBounds {
    val coordinateBounds =
      mapboxMap.coordinateBoundsForCameraUnwrapped(camera.toCameraOptions(context))
    return coordinateBounds.toFLTCoordinateBounds()
  }

  override fun coordinateBoundsZoomForCamera(camera: CameraOptions): CoordinateBoundsZoom {
    val coordinateBoundsZoom = mapboxMap.coordinateBoundsZoomForCamera(camera.toCameraOptions(context))
    return coordinateBoundsZoom.toFLTCoordinateBoundsZoom()
  }

  override fun coordinateBoundsZoomForCameraUnwrapped(camera: CameraOptions): CoordinateBoundsZoom {
    val coordinateBoundsZoom =
      mapboxMap.coordinateBoundsZoomForCameraUnwrapped(camera.toCameraOptions(context))
    return coordinateBoundsZoom.toFLTCoordinateBoundsZoom()
  }

  override fun pixelForCoordinate(coordinate: Point): ScreenCoordinate {
    val screenCoordinate = mapboxMap.pixelForCoordinate(coordinate)
    return screenCoordinate.toFLTScreenCoordinate(context)
  }

  override fun coordinateForPixel(pixel: ScreenCoordinate): Point {
    return mapboxMap.coordinateForPixel(pixel.toScreenCoordinate(context))
  }

  override fun pixelsForCoordinates(coordinates: List<Point>): MutableList<ScreenCoordinate> {
    val screenCoordinates = mapboxMap.pixelsForCoordinates(coordinates)
    return screenCoordinates.map { it.toFLTScreenCoordinate(context) }.toMutableList()
  }

  override fun coordinatesForPixels(pixels: List<ScreenCoordinate?>): List<Point> {
    return mapboxMap.coordinatesForPixels(pixels.map { it!!.toScreenCoordinate(context) })
  }

  override fun setCamera(cameraOptions: CameraOptions) {
    mapboxMap.setCamera(cameraOptions.toCameraOptions(context))
  }

  override fun getCameraState(): CameraState {
    return mapboxMap.cameraState.toCameraState(context)
  }

  override fun setBounds(options: CameraBoundsOptions) {
    mapboxMap.setBounds(options.toCameraBoundsOptions())
  }

  override fun getBounds(): CameraBounds {
    return mapboxMap.getBounds().toFLTCameraBounds()
  }
}